src/wast-lexer.cc

```diff
@@ -22,48 +22,22 @@
 #include "config.h"
 
 #include "circular-array.h"
+#include "error-handler.h"
 #include "lexer-source.h"
 #include "wast-parser.h"
-#include "wast-parser-lexer-shared.h"
-
-/* must be included after so some typedefs will be defined */
-#include "wast-parser-gen.hh"
 
 /*!max:re2c */
 
 #define INITIAL_LEXER_BUFFER_SIZE (64 * 1024)
 
-#define NAME_TO_VALUE(name) WABT_TOKEN_TYPE_##name
-
-#define LOOKAHEAD(name)                   \
-  SetLookaheadToken(NAME_TO_VALUE(name)); \
-  next_pos_ = cursor_;                    \
-  continue
-
-#define RETURN(name)                   \
-  SetToken(NAME_TO_VALUE(name));       \
-  return PopLookaheadToken(lval, loc);
-
-#define RETURN_LPAR(name)               \
-  SetToken(NAME_TO_VALUE(name));        \
-  if (IsLookaheadLpar()) {              \
-    SetLocation(loc);                   \
-    return NAME_TO_VALUE(LPAR_##name);  \
-  }                                     \
-  return PopLookaheadToken(lval, loc);
-
-#define ERROR(...)                                  \
-  SetLocation(loc);                                 \
-  WastParserError(loc, this, parser, __VA_ARGS__)
+#define ERROR(...) parser->Error(GetLocation(), __VA_ARGS__)
 
 #define BEGIN(c) cond = (c)
-#define FILL(n)                                \
-  do {                                         \
-    if (Failed(Fill(loc, parser, (n)))) {      \
-      int value = NAME_TO_VALUE(EOF);          \
-      SetToken(value);                         \
-      return PopLookaheadToken(lval, loc);     \
-    }                                          \
+#define FILL(n)              \
+  do {                       \
+    if (Failed(Fill((n)))) { \
```

Comment:
What if there were lookahead tokens? In such a case you would return eof too early.